File: day_13/p2.py
from typing import Dict, List, Tuple
import numpy as np


def parse(fname: str) -> List[Dict[str, Tuple[int, int]]]:
    lines = []
    with open(fname, 'r') as f:
        lines = [line for line in f.read().splitlines() if line]
    
    configs = []
    for i in range(0, len(lines), 3):
        ax, ay = int(''.join([char for char in lines[i].split()[-2:][0] if char.isdigit()])), \
                 int(''.join([char for char in lines[i].split()[-2:][1] if char.isdigit()]))

        bx, by = int(''.join([char for char in lines[i+1].split()[-2:][0] if char.isdigit()])), \
                 int(''.join([char for char in lines[i+1].split()[-2:][1] if char.isdigit()]))
        
        px, py = int(''.join([char for char in lines[i+2].split()[-2:][0] if char.isdigit()])), \
                 int(''.join([char for char in lines[i+2].split()[-2:][1] if char.isdigit()]))

        configs.append({'A' : (ax, ay), 'B' : (bx, by), 'prize' : (px+10000000000000, py+10000000000000)})
    
    return configs

def compute(configs: List[Dict[str, Tuple[int, int]]]) -> int:
    prices = []
    for i, config in enumerate(configs):
        coef_mat = np.matrix([config['A'], config['B']]).T
        knowns = np.matrix([[config['prize'][0]], [config['prize'][1]]])

        unknowns = np.linalg.solve(coef_mat, knowns)
        cntA, cntB = round(unknowns[0,0]), round(unknowns[1,0])

        if config['A'][0] * cntA + config['B'][0] * cntB == config['prize'][0] and \
            config['A'][1] * cntA + config['B'][1] * cntB == config['prize'][1]:
            prices.append(3 * cntA + cntB)

    return sum(prices)

def main():
    configs = parse("input.txt")
    print(compute(configs))


if __name__ == "__main__":
    main()

File: day_13/p2.cpp
#include "pypp.hpp" // https://github.com/A-Bergonzoli/pypp/tree/main

#include <Eigen/Dense>
#include <gmpxx.h>
#include <iostream>
#include <unordered_map>

using namespace std;
using Config = unordered_map<string, pair<mpz_class, mpz_class>>;
using Configs = vector<Config>;

typedef Eigen::Matrix<mpf_class, 2, 2> Matrix2_128;
typedef Eigen::Matrix<mpf_class, 2, 1> Vector2_128;

Configs Parse(const string& fname)
{
    Configs configs {};
    auto lines = pypp::splitFileLines(fname);

    for (int i { 0 }; i < lines.size(); i += 3) {
        Config config {};
        auto ax_s = pypp::splitThenGetAt(lines[i], ' ', 2);
        auto ay_s = pypp::splitThenGetAt(lines[i], ' ', 3);
        erase_if(ax_s, [](char ch) { return !isdigit(ch); });
        erase_if(ay_s, [](char ch) { return !isdigit(ch); });
        config["A"] = make_pair(mpz_class(ax_s), mpz_class(ay_s));

        auto bx_s = pypp::splitThenGetAt(lines[i + 1], ' ', 2);
        auto by_s = pypp::splitThenGetAt(lines[i + 1], ' ', 3);
        erase_if(bx_s, [](char ch) { return !isdigit(ch); });
        erase_if(by_s, [](char ch) { return !isdigit(ch); });
        config["B"] = make_pair(mpz_class(bx_s), mpz_class(by_s));

        auto px_s = pypp::splitThenGetAt(lines[i + 2], ' ', 1);
        auto py_s = pypp::splitThenGetAt(lines[i + 2], ' ', 2);
        erase_if(px_s, [](char ch) { return !isdigit(ch); });
        erase_if(py_s, [](char ch) { return !isdigit(ch); });
        config["prize"] = make_pair(mpz_class(px_s), mpz_class(py_s));
        config["prize"].first += 10000000000000UL;
        config["prize"].second += 10000000000000UL;
        configs.push_back(config);
    }

    return configs;
}

mpz_class Compute(const Configs& configs)
{
    vector<mpz_class> prices;

    for (const auto& config : configs) {
        Matrix2_128 coef_mat;
        coef_mat << config.at("A").first, config.at("B").first,
                    config.at("A").second, config.at("B").second;

        Vector2_128 knowns;
        knowns << config.at("prize").first, config.at("prize").second;

        Vector2_128 unknowns = coef_mat.inverse() * knowns;
        uint64_t cntA = round(unknowns[0].get_d());
        uint64_t cntB = round(unknowns[1].get_d());

        if (config.at("A").first * cntA + config.at("B").first * cntB == config.at("prize").first &&
            config.at("A").second * cntA + config.at("B").second * cntB == config.at("prize").second) {
            prices.push_back(3 * cntA + cntB);
        }
    }

    mpz_class total = 0;
    for (auto price : prices)
        total += price;

    return total;
}

int main()
{
    const auto configs = Parse("input.txt");
    cout << Compute(configs) << endl;

    return 0;
}
